### DISCUSSION TITLE
What happens with the "Lirc" exporter when trying to export a complex code?

### DISCUSSION THREAD

**Discussion by Arcitec** [bengtmartensson/IrScrutinizer#424] [Q&A] [answered]
Hi, I'm just curious. There's a "Lirc Raw" mode to export any code via raw timings. But I prefer the "Lirc" mode since it's more compact.

Therefore I wonder what happens in "Lirc" export mode when it encounters too-complex codes, such as the thing that was blocked in this commit:

https://github.com/bengtmartensson/IrScrutinizer/issues/363

https://github.com/bengtmartensson/IrProtocolCodeGeneration/commit/d8b124a6cf2aa077fbe45310c609d4e811f2f545

I don't think that I have such remotes at the moment. But this is just a more general question about what IrScrutinizer does during export of too-complex codes?

Does "Format:Lirc" then do a mix of compact sequences for the simple buttons, and "raw timings" sequences for the difficult buttons that it could not encode compactly?

Or does it throw away the complex buttons from the exported config completely?

Or perhaps something else? :-)

---

_PS: The more I learn about lirc thanks to your project, the less impressed I am by lirc. Wish they had accepted your ccf patches to let people easily debug / test newly captured or generated codes._ ;-) _And... let's not even talk about how bad their "irrecord" is at recording IR, holy hell..._

Thank you for creating IrScrutinizer and making the lirc ecosystem bearable. You literally saved my sanity today after I spent hours trying to record with lirc's broken tools (irrecord, mode2). :-P

**Comment by bengtmartensson** [comment, marked as answer]
> Therefore I wonder what happens in "Lirc" export mode when it encounters too-complex codes

If a protocol is not handled explicitly by the protocol transformations in exportformats.d/lirc.xml, it fallbacks to the generic transformation, which creates a raw Lirc file, lines 69--131 in the current version. (That is how templates are matched in XSLT.)

> I don't think that I have such remotes at the moment.

Just type in for example a RC5x signal in the table of Scrutinizer Remote/Parametrized, and export it.

> Does "Format:Lirc" then do a mix of compact sequences for the simple buttons, and "raw timings" sequences for the difficult buttons that it could not encode compactly?
>
> Or does it throw away the complex buttons from the exported config completely?

It generates several (Lirc) remotes in the export.

> PS: The more I learn about lirc thanks to your project, the less impressed I am by lirc. 

Yeah, the depressing thing is not that Lirc sucks, but how many people believe that Lirc is the best thing since sliced bread. See my review [here](https://sourceforge.net/projects/lirc/).

> Thank you for creating IrScrutinizer and making the lirc ecosystem bearable. You literally saved my sanity today after I spent hours trying to record with lirc's broken tools (irrecord, mode2). :-P

Thanx for those nice words! Check out [Lirconian](https://github.com/bengtmartensson/Lirconian) which is a replacement of the equally horrendous irsend.

**Comment by bengtmartensson** [comment]
@Bananaman: are you satisfied with the answer?

**Comment by Arcitec** [comment]
Hi, I'm sorry for not replying earlier. I was busy and didn't notice that you had started a discussion. I'm here now. :-)

Thank you for the fantastic answer. It's great to know that IrScrutinizer will always create valid Lirc files even when the normal non-raw mode was selected. :-)

You're honestly way beyond the skill level of the Lirc project. And it really saddens me to see people believe that Lirc is something amazing. Everything I have seen so far has been really messy and incompetent. For example, okay, let's say they don't want to let internet API users send arbitrary IR commands just in case someone sends evil commands to turn on your hairdryer in your server room? Well, then how about a local binary that has that privilege? And let's say they don't want the local binary to have that privilege? Well, then how about a config setting for people who want to allow that local access? But nope, their reply was just to shut out all the users that want a quick way to debug IR codes.

I discovered IrScrutinizer by seeing how helpful you were on their mailing list. And the stark contrast between their bad attitude and your skill was horrifying to watch.

In fact, IrScrutinizer is the only reason I was able to record my remotes, because irrecord and mode2 both failed completely.

Because of Lirc's incompetence, I had to think of workarounds and actually used IrScrutinizers "generate missing F-codes" to generate every other function, exported them all to a Lirc config, and then ran a bash script to loop through "irsend" with each of the auto-generated commands, to see which ones triggered hidden functions on my devices. I was able to find some more modes and features. All thanks to you.

You are brilliant.

Thanks for pointing me to Lirconian. I will switch to that instead if I stay on lirc. You've really cleaned up the horrendous commandline that irsend had, wow.

> See my review here.

Tried to click that but it leads to the official lirc project site. 😄 

I have one important question that arose today, actually. I'll write it in a separate message below.

**Reply by bengtmartensson** [reply]
Thanks for those nice words.

> let's say they don't want to let internet API users send arbitrary IR commands just in case someone sends evil commands to turn on your hairdryer in your server room?

See https://sourceforge.net/p/lirc/tickets/312/ ...

> Tried to click that but it leads to the official lirc project site. smile

Scroll down about two screenfulls to the "user reviews"

Lirc is a dead chicken (dead but stil running). No-one is maintaining any more. Not even a trivial, uncontroversal thing like https://sourceforge.net/p/lirc/tickets/321/ are being fixed...

**Reply by Arcitec** [reply]
Yikes, I've read the reviews and those tickets now and it just sickens me. That is a prime example of a project that needs new management, because the issue is with them, not the users or code contributors (that keep getting rejected).

**Comment by Arcitec** [comment]
I recently purchased a Irdroid USB IR Transceiver, unfortunately. Didn't realize that it was just a stolen IrToy design which had removed the original author's credits, removed the IR detector, doubled the price, removed the software reflash bootloader, and ran on a buggy old 2.0 firmware (just renamed by Irdroid to "2.5" to fool customers), with no way to upgrade without flashing hardware. Ugh. Wish I had seen your 2015 or so posts about this before I ordered. That's $40 down the drain. But at least it was a learning experience.

Your messages on the lirc mailing list about Irdroid were very entertaining. What's not so funny though is that Georgi (I think that's his name) who created the Irdroid IrToy ripoff still hasn't improved a single thing about it half a decade later and still charges over twice the IrToy price...

But after some more research, I then realized that IrToy itself is awful too. It's a messy product abandoned in 2011 and has numerous bugs (disconnects itself from host after PC sleeps due to buggy USB stack, getting stuck in debug mode by itself, and tons of IR sampling errors due to poor code).

And then I suddenly discovered your messages about something called Girs. It seems like you've created the entire Girs ecosystem yourself. And the most interesting piece was "AGirs" for a little Arduino Nano-based IR transceiver.

Would you suggest that I build that device and move over to Girs instead? Or would you suggest something else such as http://www.iguanaworks.net/products/usb-ir-transceiver/ ?

Here are my questions about AGirs:

- I was thinking of using the Arduino Nano Every but I see that it might require a different circuit and might have some issues: https://github.com/Arduino-IRremote/Arduino-IRremote/pull/657 ... Do you suggest that I pick a regular Nano clone board and use your exact circuit here instead? http://www.harctoolbox.org/arduino_nano.html
- Is AGirs able to capture and/or send at variable frequencies, or is it stuck at 38 kHz like most products?
- Is there any point in having multiple transmit LEDs, or is 1 enough? Asking because Irdroid's main "contribution" was to have 3 LEDs on the IrToy, so I am curious if that helps. Edit: I just realized your design uses 1 wide and 1 narrow beam for broad coverage. I see. Excellent!
- And most importantly: Is AGirs able to handle USB wakeup/sleep?
- I assume that the AGirs is basically a user-mode way to talk to the serial port of the Arduino, and that no kernel-mode drivers are needed for anything? Just simple "dialout" COM privilege and something like `/dev/ttyACM0` I guess?
- Is this the library that it uses internally on the Nano? https://github.com/bengtmartensson/Infrared4Arduino

I am very excited about the idea of building the AGirs IR arduino board.

My use cases are a Home Assistant server to automate various dumb IR-based home devices (so the IR dongle needs to be able to "stay alive" with minimal or no crashes/needs to unplug the dongle). And to play around with hacking devices by sending custom IR commands.

Would the AGirs Arduino IR transceiver be the path you recommend, or something else? 😄 


With best regards,

Johnny

**Reply by bengtmartensson** [reply]
> Your messages on the lirc mailing list about Irdroid were very entertaining.

Strinctly speaking, in 2021 the IrToy is not available (see http://hifi-remote.com/forums/viewtopic.php?t=102463), while the IrDroid is .... :-\

> I then realized that IrToy itself is awful too.

The non-configurable ending timeout (slightly less than 2 seconds) gets really annoying if bulk-capturing... See https://github.com/bengtmartensson/AGirs/issues/33

> Would you suggest that I build that device and move over to Girs instead?

Sure,

> Or would you suggest something else such as http://www.iguanaworks.net/products/usb-ir-transceiver/ ?

This is supported through the /dev/lirc device on Linux, since a few years. So just select /dev/lirc in IrScrutinizer.

> I was thinking of using the Arduino Nano Every but I see that it might require a different circuit and might have some issues: Arduino-IRremote/Arduino-IRremote#657 ... Do you suggest that I pick a regular Nano clone board and use your exact circuit here instead? http://www.harctoolbox.org/arduino_nano.html

Strictly speaking, the Nano does everything you need, 16MHz is enough for this application, and you cannot beat the price. The Nano Every is supported by AGirs (really Infrared4Arduino). Note that old boards like Nano and Uno sends up to 40mA (recommended 20mA), while more modern boards often are limited to just a few mA, so  you need a driver transistor with the latter (recommended: 2N7000/BS170). 

> Is AGirs able to capture and/or send at variable frequencies

Yes.

> Is AGirs able to handle USB wakeup/sleep

Short answer: I do no know. AGirs does not have any code for it; depends on hardware and low-level hardware. Probably different between different boards. Please report if you gather some experience.

> I assume that the AGirs is basically a user-mode way to talk to the serial port of the Arduino, and that no kernel-mode drivers are needed for anything

It uses a serial port in USB disguise. A kernel driver is needed for the USB chip, for example CH340. This is in general directly found by Linuxes, as well as Windows 10. Possibly a direct USB connection would be better?

> Is this the library that it uses internally on the Nano? https://github.com/bengtmartensson/Infrared4Arduino

AGirs is built upon the low level Infrared4Arduino library.

> Would the AGirs Arduino IR transceiver be the path you recommend

Absolutely. I would very much welcome more users.

**Reply by Arcitec** [reply]
Thank you so incredibly much, Bengt.

I'm definitely going to build the Arduino Nano version that you designed, because your design is really elegant with very easy assembly and flatly mounted components which will make it easy to put it in a small IRT acrylic case, and I even realized that you use both a wide and narrow beam IR LED to enhance reach. Superb.

There are some tiny remaining questions since it turns out I had misunderstood something earlier. I've marked them with numbers to make it easier to answer without quoting. :-)

> Strictly speaking, the Nano does everything you need, 16MHz is enough for this application, and you cannot beat the price. The Nano Every is supported by AGirs (really Infrared4Arduino). Note that old boards like Nano and Uno sends up to 40mA (recommended 20mA), while more modern boards often are limited to just a few mA, so you need a driver transistor with the latter (recommended: 2N7000/BS170).

Ahh, involving a transistor and all of its funky legs (spread all over the Nano board haha) would make things much more complex and it sounds like there's no reason to do that extra work.

1. Is the transistorless Nano design @ a safe 20 mA with those "Osram High Power IR LEDs" strong enough for a signal that reaches across a typical living room, ie. 6 meters range?

2. Is this bill of materials still correct? http://www.harctoolbox.org/arduino_nano.html#Appendix+A.+Bill+of+material

> Is AGirs able to capture and/or send at variable frequencies
> 
> Yes.

Oops, I thought AGirs was the name of the Arduino Nano IR Transceiver board design you made. That's the one I meant.

3. Does the hardware design at http://www.harctoolbox.org/arduino_nano.html have the ability to receive and/or transmit variable IR frequencies?

> Is AGirs able to handle USB wakeup/sleep
> 
> Short answer: I do no know. AGirs does not have any code for it; depends on hardware and low-level hardware. Probably different between different boards. Please report if you gather some experience.

I made the same mistake here. I was thinking about the hardware board itself.

4. I meant whether the hardware design is able to keep working after the PC has been in sleep mode? And I assume so since it seems to be using the Arduino's serial port interface which they've definitely made rock solid and able to handle USB sleep/resuming communication after sleep. 😄 

> It uses a serial port in USB disguise. A kernel driver is needed for the USB chip, for example CH340. This is in general directly found by Linuxes, as well as Windows 10.

Excellent! :-) And I didn't even know that Girs itself had Windows compatibility too, but just noticed that as well. Nice.

5. And the final and absolutely last question before building this: If I include the 3 debug LEDs (power, data, etc etc), can I turn them off via code or some other way so that they're disabled without having to desolder them during regular home usage?

> Absolutely. I would very much welcome more users.

Thank you for the warm welcoming. I really hope that I can contribute something back to the project when I get involved in it! Whether that's code or help for other users, time will tell! :) One thing is for sure, I will be spreading the word about switching from Lirc to Girs + IrScrutinizer.

**Reply by bengtmartensson** [reply]
1. Absolutely. The "reference configuration" blows the socks off of "all" commercial remotes, with the exception of Philips Pronto and IrTrans (these having 4 IR Leds).
2. Yes. Possibly I should add that there is a "successor" to the TSOP3*-er series by Vishay, the TSOP1* series, see https://www.vishay.com/docs/49845/sg2145-1912-infrared_receivers.pdf But the one in by list is more than good enough...
3. You can send "any" frequency, and receive "any" frequency with the non-demodulating sensor. Select another TSOP-* if you want. But it is not really used by IrScrutinizer anyhow.
4. Donno. But in general the board resets when the hosts (re-) opens the connection. (There are ways around that by cutting certain connections on the board, but I have not tried it.)
5. They can be turn off/disabled by certain commands, but that does not survive a reboot. Or you can reconfigure and recompile. (Chances are that you will find the SMD-LEDs on the board more annoying...)

BTW, be sure to read http://www.harctoolbox.org/IrScrutinizer.html#Can+I+use+this+program+for%0Aconveniently+controlling+my+favorite+IR+controlled+device+from+the+sofa%3F

**Reply by Arcitec** [reply]
Thank you so incredibly much again. I sincerely mean that.

1. Awesome, that's all I need! <3
2. Ahhh, do you know if they are drop-in replacements? `TSOP34438 ---> TSOP14438`?
3. Oh wow, so that's why it's so terrible that Irdroid removed the non-demodulating sensor from the Irtoy. That means they removed the ability to sense arbitrary timings. What a fool and a ripoff...
4. That sounds fine to me. A board resetting by itself is a lot better than a board that requires manual unplugging (like Irdroid).
5. Perfect. Great to hear that.
6. Regarding casual control: I was expecting to have to write some Home Assistant script/program that sends commands to some Girs server or runs some Girs "send_once" binary, so that's no issue. Thank you very much for pointing me to the appropriate server app.

I am super excited about joining your Girs project. 👍 😄

**Comment by bengtmartensson** [comment]
2. You should really ask Vishay, but as a principle, I think so. But be sure to read the data sheet. Note that there are two different pinouts around; if you get that wrong, "interesting" things will happen...
3. You can add a QSE-159/OPL-551 manually to the IrDroid. But I think the firmware is slighty suspect, first version(-s) of IrToy only contained a demodulating receiver, I [i]think[/i] the non-demodulating receiver is just used to get a frequency estimate, that is then "tucked onto" the measurement from the demodulating one. Since there are no different command for receiving with the different sensors...

Would writing a Girr driver for the Home assistant be a feasible project?

**Reply by Arcitec** [reply]
Thank you very much again. I feel ready to build the Nano based design soon.

2. Ah I see. I'm leaning towards picking the same one your design uses just in case. It doesn't seem like there's any major benefit to replacing that component.
3. Yikes, that's more evidence that IrToy itself was a really bad product too. Thanks for clarifying its purpose on that board.

> Would writing a Girr driver for the Home assistant be a feasible project?

Hehe, that's exactly what I am thinking about, and I am hoping to do that. But didn't want to announce or promise anything because it's possible that Home Assistant is a real cluster-fk to write plugins for. For some reason, whoever wrote their Lirc plugin didn't implement listing remotes/sending, and nobody else has extended it in the years it has existed, perhaps because it's too hard? I am mostly worried that plugins with "arbitrary lists of remotes and arbitrary sub-commands" (rather than fixed lists of static controls) could be really hard to implement. But I am absolutely going to investigate it. And if it has a sane API, I'll attempt to write a Girs plugin for Home Assistant. 😊 But if that fails, there's always the "shell_command" method in Home Assistant, which can simply call any Girs binary that in turn triggers the remote control. Looks like that's how people currently use IR transmitters with it.

> Girr 

Inspired by this? 😂 https://youtu.be/U6pQNP0Dv1s?t=665

**Reply by bengtmartensson** [reply]
> Inspired by this? joy https://youtu.be/U6pQNP0Dv1s?t=665

:grin:  

(Note the difference between the server [Girs](http://harctoolbox.org/Girs.html) and the file format [Girr](http://harctoolbox.org/Girr.html) (even though I manage to confuse it above).)

Please let me know how things turn out with Home Assistant.
